### PR TITLE
Added ClientSession to async API

### PR DESCRIPTION
**Pull request by rozza**
MongoClient, MongoDatabase and MongoCollection operations now support
being passed a ClientSession.

JAVA-2621

### CODE REVIEW

File: driver-async/src/main/com/mongodb/async/client/AggregateToCollectionThenFindOperation.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    }
+
+    public FindOperation<T> getFindOperation() {
+        return findOperation;
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public void executeAsync(final AsyncReadBinding binding, final SingleResultCallback<AsyncBatchCursor<T>> callback) {
+        aggregateToCollectionOperation.executeAsync((AsyncWriteBinding) binding, new SingleResultCallback<Void>() {
```
Comment (rozza) [inline driver-async/src/main/com/mongodb/async/client/AggregateToCollectionThenFindOperation.java]: This is controversial as we cast to an AsyncWriteBinding

File: driver-async/src/main/com/mongodb/async/client/MappingIterable.java [hunk 2]
```diff
@@ -25,21 +25,25 @@
 
 import static org.bson.assertions.Assertions.notNull;
 
-class MappingIterable<T, U> implements MongoIterable<U> {
-    private final MongoIterable<T> iterable;
-    private final Function<T, U> mapper;
+class MappingIterable<U, V> implements MongoIterable<V> {
```
Comment (rozza) [inline driver-async/src/main/com/mongodb/async/client/MappingIterable.java]: This brings it inline with the Sync one - somehow they were slightly different.

**Comment by rozza** [comment]
I opted for an async `createSession` as it keeps the errors inline with sync.

File: driver-async/src/main/com/mongodb/async/client/AggregateIterableImpl.java [hunk 3]
```diff
@@ -185,53 +140,59 @@ public void batchCursor(final SingleResultCallback<AsyncBatchCursor<TResult>> ca
...
+            FindOperation<TResult> findOperation =
+                    new FindOperation<TResult>(new MongoNamespace(namespace.getDatabaseName(), outCollection.asString().getValue()),
+                            codecRegistry.get(resultClass))
+                            .readConcern(getReadConcern())
+                            .collation(collation)
+                            .maxAwaitTime(maxAwaitTimeMS, MILLISECONDS);
+            if (getBatchSize() != null) {
+                findOperation.batchSize(getBatchSize());
             }
-            return new AwaitingWriteOperationIterable<TResult, Void>(operation, executor, delegated);
```
Comment (jyemin) [inline driver-async/src/main/com/mongodb/async/client/AggregateIterableImpl.java]: The AwaitingWriteOperationIterable class can be removed

File: driver-async/src/main/com/mongodb/async/client/AggregateToCollectionThenFindOperation.java (hunk 1 above)
Comment (jyemin) [reply driver-async/src/main/com/mongodb/async/client/AggregateToCollectionThenFindOperation.java]: I see.  I'm ok with it for release, but please open a ticket and we can address it later.  In practice it's a safe cast...

File: driver-async/src/main/com/mongodb/async/client/AggregateIterableImpl.java (hunk 3 above)
Comment (rozza) [reply driver-async/src/main/com/mongodb/async/client/AggregateIterableImpl.java]: Done.

File: driver-async/src/main/com/mongodb/async/client/AggregateIterableImpl.java [hunk 4]
```diff
@@ -185,53 +140,59 @@ public void batchCursor(final SingleResultCallback<AsyncBatchCursor<TResult>> ca
...
+            return new AggregateToCollectionThenFindOperation<TResult>(aggregateToCollectionOperation, findOperation);
         } else {
-            return new OperationIterable<TResult>(new AggregateOperation<TResult>(namespace, aggregateList, codecRegistry.get(resultClass))
```
[the remark this answers is not in the document]
Comment (rozza) [reply driver-async/src/main/com/mongodb/async/client/AggregateIterableImpl.java]: Done.

File: driver-async/src/main/com/mongodb/async/client/ClientSessionHelper.java [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+        }
+    }
+
+
+    @SuppressWarnings("deprecation")
+    void createClientSession(final ClientSessionOptions options, final SingleResultCallback<ClientSession> callback) {
+        if (mongoClient.getSettings().getCredentialList().size() > 1) {
+            callback.onResult(null, null);
+        } else {
+            ClusterDescription clusterDescription = mongoClient.getCluster().getDescription();
```
[the remark this answers is not in the document]
Comment (rozza) [reply driver-async/src/main/com/mongodb/async/client/ClientSessionHelper.java]: Done

File: driver-async/src/main/com/mongodb/async/client/ClientSessionHelper.java [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+    @SuppressWarnings("deprecation")
+    private List<ServerDescription> getServerDescriptionListToConsiderForSessionSupport(final ClusterDescription clusterDescription) {
+        if (clusterDescription.getConnectionMode() == ClusterConnectionMode.SINGLE) {
+            return clusterDescription.getAny();
+        } else {
+            return clusterDescription.getAnyPrimaryOrSecondary();
+        }
+    }
+
+    private SingleResultCallback<ServerSession> getServerSessionCallback(final ClientSessionOptions options,
```
[the remark this answers is not in the document]
Comment (rozza) [reply driver-async/src/main/com/mongodb/async/client/ClientSessionHelper.java]: Done.

File: driver-async/src/main/com/mongodb/async/client/MongoClient.java [hunk 7]
```diff
@@ -32,6 +35,18 @@
  */
 @Immutable
 public interface MongoClient extends Closeable {
+
+    /**
+     * Creates a client session.
```
[the remark this answers is not in the document]
Comment (rozza) [reply driver-async/src/main/com/mongodb/async/client/MongoClient.java]: Done.

File: driver-async/src/main/com/mongodb/async/client/MongoClient.java [hunk 8]
```diff
@@ -32,6 +35,18 @@
...
+     *
+     * @param options  the options for the client session
+     * @param callback the callback that is passed the clientSession or a {@code MongoClientException} if the MongoDB cluster to which
+     *                 this client is connected does not support sessions
+     * @mongodb.server.release 3.6
+     * @since 3.6
+     */
+    void startSession(final ClientSessionOptions options, final SingleResultCallback<ClientSession> callback);
```
[the remark this answers is not in the document]
Comment (rozza) [reply driver-async/src/main/com/mongodb/async/client/MongoClient.java]: Done.

File: driver-async/src/main/com/mongodb/async/client/MongoClientImpl.java [hunk 9]
```diff
@@ -86,23 +113,48 @@ public MongoClientSettings getSettings() {
...
+        return executeListDatabases(null, resultClass);
+    }
+
+    @Override
+    public <TResult> ListDatabasesIterable<TResult> listDatabases(final ClientSession clientSession, final Class<TResult> resultClass) {
+        notNull("clientSession", clientSession);
+        return executeListDatabases(clientSession, resultClass);
+    }
+
+    private <T> ListDatabasesIterable<T> executeListDatabases(final ClientSession clientSession, final Class<T> clazz) {
```
[the remark this answers is not in the document]
Comment (rozza) [reply driver-async/src/main/com/mongodb/async/client/MongoClientImpl.java]: Done.

File: driver-async/src/main/com/mongodb/async/client/MongoClientImpl.java [hunk 10]
```diff
@@ -112,4 +164,5 @@ Cluster getCluster() {
     ServerSessionPool getServerSessionPool() {
```
[the remark this answers is not in the document]
Comment (rozza) [reply driver-async/src/main/com/mongodb/async/client/MongoClientImpl.java]: Done.

File: driver-async/src/main/com/mongodb/async/client/MongoCollection.java [hunk 11]
```diff
@@ -404,51 +715,127 @@ void bulkWrite(List<? extends WriteModel<? extends TDocument>> requests, BulkWri
...
-     * @param options  the options to apply to the delete operation
      * @param callback the callback passed the result of the remove one operation
      * @throws com.mongodb.MongoWriteException        returned via the callback
      * @throws com.mongodb.MongoWriteConcernException returned via the callback
      * @throws com.mongodb.MongoException             returned via the callback
+     * @since 3.6
+     * @mongodb.server.release 3.6
      */
-    void deleteOne(Bson filter, DeleteOptions options, SingleResultCallback<DeleteResult> callback);
+    void deleteOne(ClientSession clientSession, Bson filter, SingleResultCallback<DeleteResult> callback);
```
[the remark this answers is not in the document]
Comment (rozza) [reply driver-async/src/main/com/mongodb/async/client/MongoCollection.java]: Done.

File: driver-async/src/main/com/mongodb/async/client/MongoDatabaseImpl.java [hunk 12]
```diff
@@ -111,23 +112,50 @@ public MongoDatabase withReadConcern(final ReadConcern readConcern) {
...
+        return executeListCollections(null, resultClass);
+    }
+
+    @Override
+    public <TResult> ListCollectionsIterable<TResult> listCollections(final ClientSession clientSession, final Class<TResult> resultClass) {
+        notNull("clientSession", clientSession);
+        return executeListCollections(clientSession, resultClass);
+    }
+
+    private <TResult> ListCollectionsIterable<TResult> executeListCollections(final ClientSession clientSession,
```
[the remark this answers is not in the document]
Comment (rozza) [reply driver-async/src/main/com/mongodb/async/client/MongoDatabaseImpl.java]: Done.

**Comment by rozza** [comment]
Patch build in process: https://evergreen.mongodb.com/version/5a1575d1c9ec446a1c002930

File: driver-async/src/main/com/mongodb/async/client/AggregateToCollectionThenFindOperation.java (hunk 1 above)
Comment (rozza) [reply driver-async/src/main/com/mongodb/async/client/AggregateToCollectionThenFindOperation.java]: Added JAVA-2673

File: driver-async/src/main/com/mongodb/async/client/MongoClient.java [hunk 13]
```diff
@@ -39,13 +39,15 @@
     /**
      * Creates a client session.
      *
+     * <p>Note: ClientSessions cannot be used concurrently, only one can be used at a time.</p>
```
Comment (jyemin) [inline driver-async/src/main/com/mongodb/async/client/MongoClient.java]: This isn't quite right.  More than one can be used at a time, but one can't be used concurrently in multiple operations.  How about:

     A ClientSession instance can not be used concurrently in multiple asynchronous operations.

File: driver-core/src/main/com/mongodb/connection/Cluster.java [hunk 14]
```diff
@@ -45,6 +45,13 @@
      */
     ClusterDescription getDescription();
 
+    /**
+     * Get the current description of this cluster.
+     *
+     * @return the current ClusterDescription representing the current state of the cluster or null
```
Comment (jyemin) [inline driver-core/src/main/com/mongodb/connection/Cluster.java]: This shouldn't be allowed to return null and in practice it does not.  Both subclasses of BaseCluster ensure that it's initialized in their respective constructors

Comment (rozza) [reply driver-core/src/main/com/mongodb/connection/Cluster.java]: Done.

File: driver-async/src/main/com/mongodb/async/client/MongoClient.java (hunk 13 above)
Comment (rozza) [reply driver-async/src/main/com/mongodb/async/client/MongoClient.java]: Done.

**Pull request closed without merging**